opm/autodiff/NewtonIterationBlackoilCPR.cpp

```diff
@@ -188,11 +189,14 @@ namespace Opm
             const ParallelISTLInformation& info =
                 boost::any_cast<const ParallelISTLInformation&>( parallelInformation_);
             Comm istlComm(info.communicator());
-            info.copyValuesTo(istlComm.indexSet(), istlComm.remoteIndices());
+            Comm istlAeComm(info.communicator());
+            info.copyValuesTo(istlAeComm.indexSet(), istlAeComm.remoteIndices());
+            info.copyValuesTo(istlComm.indexSet(), istlComm.remoteIndices(),
+                              istlAe.N(), istlA.N()/istlAe.N());
```

Comment:
On Wed, May 20, 2015 at 03:51:49AM -0700, Atgeirr Flø Rasmussen wrote:
```> > @@ -188,11 +189,14 @@ namespace Opm
> >              const ParallelISTLInformation& info =
> >                  boost::any_cast<const ParallelISTLInformation&>( parallelInformation_);
> >              Comm istlComm(info.communicator());
> > -            info.copyValuesTo(istlComm.indexSet(), istlComm.remoteIndices());
> > +            Comm istlAeComm(info.communicator());
> > +            info.copyValuesTo(istlAeComm.indexSet(), istlAeComm.remoteIndices());
> > +            info.copyValuesTo(istlComm.indexSet(), istlComm.remoteIndices(),
> > +                              istlAe.N(), istlA.N()/istlAe.N());
> 
> I have some trouble understanding what really happens in the preceding four lines. Could you try to explain?```

info contains the parallel index set and the remote index information
for the cells of the grid. I the CPR preconditioner the iterative
solvers deal with two linear systems: The whole one consisting of
three phases, i.e. three degrees of freedom per cell. And just the
elliptic part with one degree of freedom per cell.

For the whole system we need to communicate three unknowns per cell to
get correct results. Do this with a ```BCRSMatrix<FieldMatrix<double,1,1> >```
we need one parallel index per unknown. Such an index set is construct
with the last two lines. It takes the existing index set with global
indices (i_0, ..., i_n) and creates a new one (i_0, ..., i_n,
(i_n+1)+i_0, ..., (i_n+1)+i_n, 2*(i_n+1)+i_0, ..., 2*(i_n+1)+i_n) and
sets up all the communication infrastructure.
The line above that does this for the ellipitc part, global indices
i_0, ..., i_n).